token/program-2022/src/state.rs

```diff
@@ -80,6 +103,55 @@ impl Pack for Mint {
     }
 }
 
+/// MintWithTransferFee data.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct MintWithTransferFee {
+    /// Base mint data, done as composition
+    pub base_mint: Mint,
+
+    /// ALL NEW DATA MUST START FROM HERE
+    /// account type, always set to "MintWithTransferFee"
+    pub account_type: AccountType,
+    /// Any provided fee account must be owned by this key. If set to `None`, then
+    /// this mint behaves like a normal `Mint`.
+    pub fee_account_owner: COption<Pubkey>,
+    /// Amount of transfer collected as fees, expressed as basis points of the
+    /// transfer amount, ie.  increments of 0.01%
+    pub transfer_fee_basis_points: u16,
+    /// Maximum fee assessed on transfers, expressed as an amount of tokens
+    pub maximum_fee: u64,
+}
+
+/// `UberMint` contains all of the possible data in any Mint, to be used instead
+/// of `Mint` in programs.  Instead of using `Mint::unpack`, use
+/// `UberMint::unpack`, and the unpack works for any Mint type, filling in
+/// the appropriate type.
+/// When writing data back to the buffer, `UberMint::pack` respects the type contained,
+/// and errors if unusable fields have been populated or changed.  For example,
+/// if the fee is set on a normal Mint, it will error.
+///
+/// Name TBD. Other options: MetaMint, AllMint, OverMint, MegaMint.
+/// OR we rename Mint to BaseMint, and then call this Mint.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct UberMint {
```

Comment:
I'm not yet convinced about the utility of the uber structs, I suspect it'll be more flexible to make the caller deal with the different types individually because it may not always be the case that new account types will be a superset of older ones.  But happy to save this chat for later too

Reply:
That's exactly the idea! As you say, if `Mint3` does not contain all of the fields in `Mint2`, then `UberMint` has all the fields for Mint3 and Mint2, with sensible defaults.  This means a defi protocol only ever has to deal with `UberMint` to do its work.  Either way, you're right, we don't have to figure this out now since there's no `Mint3`.

It may be worth having conversations with teams to see what works best for them. I'm thinking through the case of a defi protocol saying, "ok cool, new mint type. What do I need to do to support it?" I'm seeing two main approaches:

* have an `unpack` function that returns a `Mint` regardless of what's inside, and another that returns a `Mint2` regardless of what's inside
* have an `unpack` function that returns a trait object / superstruct

Happy to hear any other ideas!

Reply:
I was thinking of essentially the `From` and `TryFrom` traits.  If one wants a specific `AccountType`, they use `From`.  `TryFrom` can be used if the actual `AccountType` can be converted to the desired type in a sensible way.

But I feel like we don't need to hammer out this detail yet, and I agree that discussing with external teams would be valuable

Reply:
Sounds good to me. I'll remove these structs in the meantime. I put them in to kickstart the discussion